libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -672,20 +672,37 @@ string YulUtilFunctions::overflowCheckedIntMulFunction(IntegerType const& _type)
 			function <functionName>(x, y) -> product {
 				x := <cleanupFunction>(x)
 				y := <cleanupFunction>(y)
+				product := mul(x, y)
 				<?signed>
-					// overflow, if x > 0, y > 0 and x > (maxValue / y)
-					if and(and(sgt(x, 0), sgt(y, 0)), gt(x, div(<maxValue>, y))) { <panic>() }
-					// underflow, if x > 0, y < 0 and y < (minValue / x)
-					if and(and(sgt(x, 0), slt(y, 0)), slt(y, sdiv(<minValue>, x))) { <panic>() }
-					// underflow, if x < 0, y > 0 and x < (minValue / y)
-					if and(and(slt(x, 0), sgt(y, 0)), slt(x, sdiv(<minValue>, y))) { <panic>() }
-					// overflow, if x < 0, y < 0 and x < (maxValue / y)
-					if and(and(slt(x, 0), slt(y, 0)), slt(x, sdiv(<maxValue>, y))) { <panic>() }
+					<?gt128bit>
+						// overflow, if x > 0, y > 0 and x > (maxValue / y)
+						if and(and(sgt(x, 0), sgt(y, 0)), gt(x, div(<maxValue>, y))) { <panic>() }
+						// underflow, if x > 0, y < 0 and y < (minValue / x)
+						if and(and(sgt(x, 0), slt(y, 0)), slt(y, sdiv(<minValue>, x))) { <panic>() }
+						// underflow, if x < 0, y > 0 and x < (minValue / y)
+						if and(and(slt(x, 0), sgt(y, 0)), slt(x, sdiv(<minValue>, y))) { <panic>() }
+						// overflow, if x < 0, y < 0 and x < (maxValue / y)
+						if and(and(slt(x, 0), slt(y, 0)), slt(x, sdiv(<maxValue>, y))) { <panic>() }
+					<!gt128bit>
+						// overflow, if same signal and product > maxValue
+						if and(
+							iszero(and(xor(x,y), <bitMask>)),
```

Comment:
Isn't it just enough to check `if or(sgt(product, <maxValue>), slt(product, <minValue>)) { <panic>() }" for this case?

Reply:
Yes, it makes more sense.